node/network/availability-recovery/src/lib.rs

```diff
@@ -0,0 +1,756 @@
+			tracing::warn!(
+				target: LOG_TARGET,
+				err = ?e,
+				"Error responding with an availability recovery result",
+			);
+		}
+		return Ok(());
+	}
+
+	if let Some(interaction) = state.interactions.get_mut(&candidate_hash) {
+		interaction.awaiting.push(response_sender);
+		return Ok(());
+	}
+
+	let session_index = request_session_index_for_child_ctx(state.live_block_hash, ctx).await?.await??;
+	let session_info = request_session_info_ctx(state.live_block_hash, session_index, ctx).await?.await??;
+
+	match session_info {
+		Some(session_info) => {
+			launch_interaction(
+				state,
+				ctx,
+				session_index,
+				session_info,
+				receipt,
+				response_sender,
+			).await
+		}
+		None => {
+			tracing::warn!(
+				target: LOG_TARGET,
+				"SessionInfo is `None` at {}", state.live_block_hash,
+			);
+			Ok(())
+		}
+	}
+}
+
+/// Queries a chunk from av-store.
+#[tracing::instrument(level = "trace", skip(ctx), fields(subsystem = LOG_TARGET))]
+async fn query_chunk(
+	ctx: &mut impl SubsystemContext<Message = AvailabilityRecoveryMessage>,
+	candidate_hash: CandidateHash,
+	validator_index: ValidatorIndex,
+) -> error::Result<Option<ErasureChunk>> {
+	let (tx, rx) = oneshot::channel();
+	ctx.send_message(AllMessages::AvailabilityStore(
+		AvailabilityStoreMessage::QueryChunk(candidate_hash, validator_index, tx),
+	)).await;
+
+	Ok(rx.await?)
+}
+
+/// Handles message from interaction.
+#[tracing::instrument(level = "trace", skip(ctx, state), fields(subsystem = LOG_TARGET))]
+async fn handle_from_interaction(
+	state: &mut State,
+	ctx: &mut impl SubsystemContext<Message = AvailabilityRecoveryMessage>,
+	from_interaction: FromInteraction,
+) -> error::Result<()> {
+	match from_interaction {
+		FromInteraction::Concluded(candidate_hash, result) => {
+			// Load the entry from the interactions map.
+			// It should always exist, if not for logic errors.
+			if let Some(interaction) = state.interactions.remove(&candidate_hash) {
+				// Send the result to each member of awaiting.
+				for awaiting in interaction.awaiting {
+					awaiting.send(result.clone()).map_err(|_| oneshot::Canceled)?;
+				}
+			} else {
+				tracing::warn!(
+					target: LOG_TARGET,
+					"Interaction under candidate hash {} is missing",
+					candidate_hash,
+				);
+			}
+
+			state.availability_lru.put(candidate_hash, result);
+		}
+		FromInteraction::MakeRequest(id, candidate_hash, validator_index, response) => {
+			let relay_parent = state.live_block_hash;
+
+			// Take the validators we are already connecting to and merge them with the
+			// newly requested one to create a new connection request.
+			let new_discovering_validators_set = state.discovering_validators.keys()
+				.cloned()
+				.chain(std::iter::once(id.clone()))
+				.collect();
+
+			// Issue a NetworkBridgeMessage::ConnectToValidators.
+			// Add the stream of connected validator events to state.connecting_validators.
+			match validator_discovery::connect_to_validators(
```

Comment:
Should use `connect_to_past_session_validators`, I believe

Reply:
it's a little bit of a misnomer because it also works in the current session, but we need to get data for past sessions as well.

Reply:
I believe that it just computes the necessary session index every time, so it would compute the past session for `relay_parent` if needed.

https://github.com/paritytech/polkadot/blob/2c6920c40fb5ffb607cd2d3607dd182a76cee5d3/node/subsystem-util/src/validator_discovery.rs#L40-L47

Reply:
Yes, but we're using `relay_parent=state.live_block_hash` here, which is some arbitrary recent block we believe the state to still be present for. It's not the relay-parent of the candidate we're requesting, which we shouldn't use anyway because the state may not be available.

So my original comment is still correct